src/Kvasir/Reconstitution/Lookups.cs

```diff
@@ -30,19 +30,30 @@ internal static partial class Lookup {
         public static object ByKey(DBData key, IEnumerable<object> entities, DataExtractionPlan extractor) {
             Debug.Assert(entities.All(e => e.GetType().IsInstanceOf(extractor.ExpectedSource)));
 
+            if (CacheDictionary.ContainsKey(key))
+                return CacheDictionary[key];
+
             foreach (var entity in entities) {
+                if (CacheValues.Contains(entity))
+                    continue;
                 var keyIdx = 0;
                 foreach (var value in extractor.ExecutePiecewise(entity)) {
                     if (!value.Equals(key[keyIdx++])) {
                         goto nextEntity;
                     }
                 }
+                CacheDictionary[key] = entity;
+                CacheValues.Add(entity);
                 return entity;
                 nextEntity: continue;
             }
 
             // This code should be unreachable
             throw new ApplicationException($"No entity found to match key: {string.Join(", ", key)}");
         }
+
```

Comment:
I like to put two blank lines between my functions and my class state, it helps visually identify the separation between the API (properties, constructor, methods) and the internal variables